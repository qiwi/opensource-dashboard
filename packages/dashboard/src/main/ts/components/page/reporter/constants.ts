export const treeDefault = {
  type: 'tree',
  depType: 'default',
  packageNamePattern: '',
  source: 'auto',
}

export const versionsDefault = {
  type: 'versions',
  depType: 'default',
  packageNamePattern: '',
  source: 'auto',
  versionRange: '',
}

export const usageDefault = {
  type: 'usage',
  depType: 'default',
  packageNamePattern: '',
  source: 'auto',
  versionRange: '',
}

export const defaultTypes = {
  tree: treeDefault,
  usage: usageDefault,
  versions: versionsDefault,
}
