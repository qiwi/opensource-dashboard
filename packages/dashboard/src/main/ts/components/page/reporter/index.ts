export {
  versionsDefault,
  treeDefault,
  usageDefault,
  defaultTypes,
} from './constants'
export { Reporter } from './reporter'
