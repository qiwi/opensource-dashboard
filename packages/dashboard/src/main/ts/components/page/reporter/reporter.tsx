import { Block, Card, Flex, Box } from '@qiwi/pijma-core'
import { Button, HeaderMenu } from '@qiwi/pijma-desktop'
import React, { useReducer } from 'react'
import { reportService } from '../../../service'
import { useAsync } from '../../../hooks/useAsyncExecutor'

import {
  TreeReportForm,
  VersionsReportForm,
  UsageReportForm,
  TReportProps,
} from '../../org/reportForm'
import { defaultTypes, treeDefault } from './constants'

function downloadObjectAsJson(exportObj: any, exportName: string) {
  const dataStr =
    'data:text/json;charset=utf-8,' +
    encodeURIComponent(JSON.stringify(exportObj))
  const downloadAnchorNode = document.createElement('a')
  downloadAnchorNode.setAttribute('href', dataStr)
  downloadAnchorNode.setAttribute('download', exportName + '.json')
  document.body.appendChild(downloadAnchorNode) // required for firefox
  downloadAnchorNode.click()
  downloadAnchorNode.remove()
}

const reducer = (
  state: Record<string, string>,
  action: Record<string, string>,
) => {
  if (action.type === 'type') {
    return defaultTypes[action.data as 'usage' | 'versions' | 'tree']
  }

  return { ...state, [action.type]: action.data }
}

const getReport = async ({
  type,
  depType,
  packageNamePattern,
  source,
  versionRange,
}: any) => {
  versionRange = versionRange || undefined
  const result = await reportService.getReport(type, {
    depType,
    packageNamePattern,
    source,
    versionRange,
  })

  downloadObjectAsJson(result, `${type}Report`)
}

export const Reporter = () => {
  const [data, setData] = useReducer(reducer, treeDefault)
  const [execute, status] = useAsync(getReport)
  return (
    <Flex mt={6}>
      <Block css={{ width: '100%', padding: '48px' }}>
        <Card width={1} height={14} mb={'32px'}>
          <HeaderMenu>
            {[
              {
                title: 'tree',
                active: data.type === 'tree',
                onClick: () => setData({ type: 'type', data: 'tree' }),
              },
              {
                title: 'usage',
                active: data.type === 'usage',
                onClick: () => setData({ type: 'type', data: 'usage' }),
              },
              {
                title: 'versions',
                active: data.type === 'versions',
                onClick: () => setData({ type: 'type', data: 'versions' }),
              },
            ]}
          </HeaderMenu>
        </Card>
        {
          {
            tree: (
              <TreeReportForm
                data={data as TReportProps['data']}
                setData={setData}
              />
            ),
            usage: (
              <UsageReportForm
                data={data as TReportProps['data']}
                setData={setData}
              />
            ),
            versions: (
              <VersionsReportForm
                data={data as TReportProps['data']}
                setData={setData}
              />
            ),
          }[data.type as 'usage' | 'versions' | 'tree']
        }
        <Box width={64}>
          <Button
            loading={status === 'pending'}
            disabled={status === 'pending'}
            type="button"
            kind="brand"
            size="normal"
            text="Получить отчет"
            onClick={() => execute(data)}
          />
        </Box>
      </Block>
    </Flex>
  )
}
