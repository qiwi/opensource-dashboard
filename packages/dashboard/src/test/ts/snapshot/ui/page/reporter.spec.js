import React from 'react'
import { PijmaContext } from '../../../pijmaContext'
import { render } from '@testing-library/react'
import { Reporter } from '../../../../../main/ts/components/page/reporter/reporter'
import '@testing-library/jest-dom/extend-expect'

describe('Reporter', () => {
  it('renders correctly', () => {
    const { container } = render(
      <PijmaContext>
        <Reporter />
      </PijmaContext>,
    )
    expect(container).toMatchSnapshot()
  })
})
