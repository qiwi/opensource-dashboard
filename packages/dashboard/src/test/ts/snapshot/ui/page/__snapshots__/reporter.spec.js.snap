// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reporter renders correctly 1`] = `
<div>
  <div
    class="css-1rwr66l"
  >
    <div
      class="css-4gfk9w"
    >
      <div
        class="css-gxgrlf"
      >
        <div
          class="css-1ns1pgm"
        >
          <div
            class="css-13o7eu2"
          >
            <div
              class="css-fqzivi"
            >
              <div
                class="css-1hpx0py"
              >
                <span
                  class="css-l3al1k"
                >
                  tree
                </span>
              </div>
              <div
                class="css-1pbsily"
              >
                <div
                  class="css-n2k0rv"
                />
              </div>
            </div>
          </div>
          <div
            class="css-1tdlxdt"
          >
            <div
              class="css-fqzivi"
            >
              <div
                class="css-1hpx0py"
              >
                <span
                  class="css-l3al1k"
                >
                  usage
                </span>
              </div>
            </div>
          </div>
          <div
            class="css-1tdlxdt"
          >
            <div
              class="css-fqzivi"
            >
              <div
                class="css-1hpx0py"
              >
                <span
                  class="css-l3al1k"
                >
                  versions
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="css-10ic8ni"
        >
          <div
            class="css-1haykwu"
          >
            <div
              class="css-1hy2az2"
            >
              <div
                class="css-dlnwt1"
              >
                <label
                  class="css-195rf93"
                >
                  packageNamePattern
                </label>
              </div>
            </div>
            <div
              class="css-1bx5ylf"
            >
              <div
                class="css-70qvj9"
              >
                <input
                  autocomplete="off"
                  class="css-amr1ku"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="css-1abbrc"
            />
          </div>
        </div>
        <div
          class="css-adxqe4"
        >
          <div
            class="css-13o7eu2"
          >
            <div
              class="css-2stwun"
            >
              <div
                class="css-15rwwo"
              >
                <label
                  class="css-1yifgsp"
                >
                  depType
                </label>
              </div>
            </div>
            <div
              class="css-tprhxn"
            >
              <div
                class="css-13o7eu2"
                tabindex="0"
              >
                <div
                  class="css-2stwun"
                >
                  <div
                    class="css-z0xd9q"
                  >
                    <div
                      class="css-196huyw"
                    >
                      <div
                        class="css-1mss6i9"
                      >
                        <div
                          class="css-110z6b4"
                        >
                          <svg
                            class="css-nmzr3v"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <rect
                              class="css-qz95lw"
                              fill="#e6e6e6"
                              height="18"
                              rx="9"
                              ry="9"
                              width="18"
                              x="3"
                              y="3"
                            />
                            <path
                              class="css-qz95lw"
                              d="M12 3a9 9 0 1 1 0 18 9 9 0 0 1 0-18zm0 2a7 7 0 1 0 0 14 7 7 0 0 0 0-14z"
                              fill="transparent"
                            />
                            <rect
                              class="css-qz95lw"
                              fill="#000"
                              height="8"
                              rx="4"
                              ry="4"
                              width="8"
                              x="8"
                              y="8"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="css-j7qwjs"
                      >
                        <div
                          class="css-59tnyu"
                        >
                          default
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="css-18ifh88"
                >
                  <div
                    class="css-z0xd9q"
                  >
                    <div
                      class="css-196huyw"
                    >
                      <div
                        class="css-1mss6i9"
                      >
                        <div
                          class="css-110z6b4"
                        >
                          <svg
                            class="css-nmzr3v"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <rect
                              class="css-qz95lw"
                              fill="#e6e6e6"
                              height="18"
                              rx="9"
                              ry="9"
                              width="18"
                              x="3"
                              y="3"
                            />
                            <path
                              class="css-qz95lw"
                              d="M12 3a9 9 0 1 1 0 18 9 9 0 0 1 0-18zm0 2a7 7 0 1 0 0 14 7 7 0 0 0 0-14z"
                              fill="transparent"
                            />
                            <rect
                              class="css-qz95lw"
                              fill="transparent"
                              height="8"
                              rx="4"
                              ry="4"
                              width="8"
                              x="8"
                              y="8"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="css-j7qwjs"
                      >
                        <div
                          class="css-59tnyu"
                        >
                          optional
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="css-18ifh88"
                >
                  <div
                    class="css-z0xd9q"
                  >
                    <div
                      class="css-196huyw"
                    >
                      <div
                        class="css-1mss6i9"
                      >
                        <div
                          class="css-110z6b4"
                        >
                          <svg
                            class="css-nmzr3v"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <rect
                              class="css-qz95lw"
                              fill="#e6e6e6"
                              height="18"
                              rx="9"
                              ry="9"
                              width="18"
                              x="3"
                              y="3"
                            />
                            <path
                              class="css-qz95lw"
                              d="M12 3a9 9 0 1 1 0 18 9 9 0 0 1 0-18zm0 2a7 7 0 1 0 0 14 7 7 0 0 0 0-14z"
                              fill="transparent"
                            />
                            <rect
                              class="css-qz95lw"
                              fill="transparent"
                              height="8"
                              rx="4"
                              ry="4"
                              width="8"
                              x="8"
                              y="8"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="css-j7qwjs"
                      >
                        <div
                          class="css-59tnyu"
                        >
                          peer
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="css-18ifh88"
                >
                  <div
                    class="css-z0xd9q"
                  >
                    <div
                      class="css-196huyw"
                    >
                      <div
                        class="css-1mss6i9"
                      >
                        <div
                          class="css-110z6b4"
                        >
                          <svg
                            class="css-nmzr3v"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <rect
                              class="css-qz95lw"
                              fill="#e6e6e6"
                              height="18"
                              rx="9"
                              ry="9"
                              width="18"
                              x="3"
                              y="3"
                            />
                            <path
                              class="css-qz95lw"
                              d="M12 3a9 9 0 1 1 0 18 9 9 0 0 1 0-18zm0 2a7 7 0 1 0 0 14 7 7 0 0 0 0-14z"
                              fill="transparent"
                            />
                            <rect
                              class="css-qz95lw"
                              fill="transparent"
                              height="8"
                              rx="4"
                              ry="4"
                              width="8"
                              x="8"
                              y="8"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="css-j7qwjs"
                      >
                        <div
                          class="css-59tnyu"
                        >
                          dev
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="css-tprhxn"
                >
                  <div
                    class="css-z0xd9q"
                  >
                    <div
                      class="css-196huyw"
                    >
                      <div
                        class="css-1mss6i9"
                      >
                        <div
                          class="css-110z6b4"
                        >
                          <svg
                            class="css-nmzr3v"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <rect
                              class="css-qz95lw"
                              fill="#e6e6e6"
                              height="18"
                              rx="9"
                              ry="9"
                              width="18"
                              x="3"
                              y="3"
                            />
                            <path
                              class="css-qz95lw"
                              d="M12 3a9 9 0 1 1 0 18 9 9 0 0 1 0-18zm0 2a7 7 0 1 0 0 14 7 7 0 0 0 0-14z"
                              fill="transparent"
                            />
                            <rect
                              class="css-qz95lw"
                              fill="transparent"
                              height="8"
                              rx="4"
                              ry="4"
                              width="8"
                              x="8"
                              y="8"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="css-j7qwjs"
                      >
                        <div
                          class="css-59tnyu"
                        >
                          all
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-adxqe4"
        >
          <div
            class="css-13o7eu2"
          >
            <div
              class="css-2stwun"
            >
              <div
                class="css-15rwwo"
              >
                <label
                  class="css-1yifgsp"
                >
                  source
                </label>
              </div>
            </div>
            <div
              class="css-tprhxn"
            >
              <div
                class="css-13o7eu2"
                tabindex="0"
              >
                <div
                  class="css-2stwun"
                >
                  <div
                    class="css-z0xd9q"
                  >
                    <div
                      class="css-196huyw"
                    >
                      <div
                        class="css-1mss6i9"
                      >
                        <div
                          class="css-110z6b4"
                        >
                          <svg
                            class="css-nmzr3v"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <rect
                              class="css-qz95lw"
                              fill="#e6e6e6"
                              height="18"
                              rx="9"
                              ry="9"
                              width="18"
                              x="3"
                              y="3"
                            />
                            <path
                              class="css-qz95lw"
                              d="M12 3a9 9 0 1 1 0 18 9 9 0 0 1 0-18zm0 2a7 7 0 1 0 0 14 7 7 0 0 0 0-14z"
                              fill="transparent"
                            />
                            <rect
                              class="css-qz95lw"
                              fill="#000"
                              height="8"
                              rx="4"
                              ry="4"
                              width="8"
                              x="8"
                              y="8"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="css-j7qwjs"
                      >
                        <div
                          class="css-59tnyu"
                        >
                          auto
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="css-18ifh88"
                >
                  <div
                    class="css-z0xd9q"
                  >
                    <div
                      class="css-196huyw"
                    >
                      <div
                        class="css-1mss6i9"
                      >
                        <div
                          class="css-110z6b4"
                        >
                          <svg
                            class="css-nmzr3v"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <rect
                              class="css-qz95lw"
                              fill="#e6e6e6"
                              height="18"
                              rx="9"
                              ry="9"
                              width="18"
                              x="3"
                              y="3"
                            />
                            <path
                              class="css-qz95lw"
                              d="M12 3a9 9 0 1 1 0 18 9 9 0 0 1 0-18zm0 2a7 7 0 1 0 0 14 7 7 0 0 0 0-14z"
                              fill="transparent"
                            />
                            <rect
                              class="css-qz95lw"
                              fill="transparent"
                              height="8"
                              rx="4"
                              ry="4"
                              width="8"
                              x="8"
                              y="8"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="css-j7qwjs"
                      >
                        <div
                          class="css-59tnyu"
                        >
                          lock
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="css-tprhxn"
                >
                  <div
                    class="css-z0xd9q"
                  >
                    <div
                      class="css-196huyw"
                    >
                      <div
                        class="css-1mss6i9"
                      >
                        <div
                          class="css-110z6b4"
                        >
                          <svg
                            class="css-nmzr3v"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <rect
                              class="css-qz95lw"
                              fill="#e6e6e6"
                              height="18"
                              rx="9"
                              ry="9"
                              width="18"
                              x="3"
                              y="3"
                            />
                            <path
                              class="css-qz95lw"
                              d="M12 3a9 9 0 1 1 0 18 9 9 0 0 1 0-18zm0 2a7 7 0 1 0 0 14 7 7 0 0 0 0-14z"
                              fill="transparent"
                            />
                            <rect
                              class="css-qz95lw"
                              fill="transparent"
                              height="8"
                              rx="4"
                              ry="4"
                              width="8"
                              x="8"
                              y="8"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="css-j7qwjs"
                      >
                        <div
                          class="css-59tnyu"
                        >
                          package
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-1v8vc71"
      >
        <button
          class="css-1mzvba8"
          type="button"
        >
          <div
            class="css-bbfvd9"
          >
            <div
              class="css-1comd0l"
            >
              <div
                class="css-y9v9ql"
              >
                <div
                  class="css-1r1bp18"
                >
                  <svg
                    class="css-z0ext1"
                    focusable="false"
                    viewBox="24 24 48 48"
                  >
                    <circle
                      class="css-1i0f4sf"
                      cx="48"
                      cy="48"
                      fill="none"
                      r="20"
                      stroke="#fff"
                      stroke-dasharray="89, 200"
                      stroke-dashoffset="0"
                      stroke-linecap="round"
                      stroke-width="4"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="css-1xx5zlq"
              >
                <div
                  class="css-1yr96g1"
                >
                  Получить отчет
                </div>
              </div>
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
